h/h_api/model/base.py

```diff
@@ -6,12 +6,37 @@ class Model:
     the responsibility of the implementer.
     """
 
-    def __init__(self, raw):
+    # An optional schema to apply to the contents when set
+    validator = None
+
+    # Custom message to add to any validation errors
+    validation_error_title = None
+
+    def __init__(self, raw, validate=True):
         """
         :param raw: The raw data to add to this object
         """
         self.raw = raw
 
+        if validate and raw is not None:
+            self.validate()
+
+    def validate(self, error_title=None):
+        """
+        Validate the contents of this object against the schema (if any)
+
+        If `validation_error_title` is set, then this will be used as a default
+        `error_title`.
+
+        :param error_title: A custom error message when errors are found
+        :raise SchemaValidationError: When errors are found
+        """
+        if error_title is None:
+            error_title = self.validation_error_title
+
+        if self.validator is not None:
+            self.validator.validate_all(self.raw, error_title)
```

Comment:
So here we have validation being applied to each different sub-part of the request (each model instance) separately, we'll presumably be calling this `Model.validate()` for each model object that makes up the whole request

Reply:
This is applying a single schema to this model. This happens during `__init__` for the object, which is always indirectly called if you go through `create()`.

So each model will validate itself if a validator is defined. This always happens on instantiation but is available on demand. Whether this happens to each part depends on whether you set the validator + whether you went through the model object.

It's possible to have the same data checked a few times if you have nested model objects which all have validators and are used in each others create methods. The nice thing is the one at the bottom will raise. They can also have different concerns. For example the `JSONAPIData` model applies a general model, which the others can assume.

I think `validate_all` is a bit misleading. `DraftV7Validator` objects have a `validate` method which raises at the first error and an `iter_errors` which returns all of them. This is a kind of a mixture which iterates all of the errors and raises a combined exception. I'll see if I can come up with a better name.